### PR TITLE
Mechanism to reduce redundancy in Maven profiles for shims [databricks]

### PR DESCRIPTION
**Pull request by gerashegalov**
Replace multiple add-source executions with a single add-source execution and a global execution computing all source sets in one pass.
- proposes pattern reuse rules
- removes a duplicate definition ShimSupportsRuntimeFiltering in 330cdh unnoticed because that shim was including 320+/scala but not 320+/java

Signed-off-by: Gera Shegalov <gera@apache.org>

### CODE REVIEW

File: sql-plugin/pom.xml [hunk 1]
```diff
@@ -460,6 +460,125 @@
...
+                            </pathconvert>
+                            <pathconvert property="spark320.sources" pathsep=",">
+                                <dirset dir="${project.basedir}/src/main">
+                                    <include name="320/*"/>
+                                </dirset>
+                            </pathconvert>
+                            <pathconvert property="spark321X.sources" pathsep=",">
+                                <dirset dir="${project.basedir}/src/main">
+                                    <include name="311+-nondb/*"/>
+                                    <include name="311until330*/*"/>
```
Comment (tgravescs) [inline sql-plugin/pom.xml:516]: isn't this spark320X.sources + some other?

File: sql-plugin/pom.xml [hunk 2]
```diff
@@ -460,6 +460,125 @@
...
+                        <exportAntProperties>true</exportAntProperties>
+                        <target>
+                            <pathconvert property="spark31X.sources" pathsep=",">
+                                <dirset dir="${project.basedir}/src/main">
+                                    <include name="311+-nondb/*"/>
+                                    <include name="311until*/*"/>
+                                    <include name="pre320-treenode/*"/>
+                                </dirset>
+                            </pathconvert>
+                            <!-- TODO loop? -->
```
[the remark this answers is not in the document]
Comment (gerashegalov) [reply sql-plugin/pom.xml:477]: could do a for loop once it's clear that this method works

File: pom.xml [hunk 3]
```diff
@@ -694,38 +520,6 @@
                 <rat.consoleOutput>true</rat.consoleOutput>
                 <parquet.hadoop.version>1.12.0</parquet.hadoop.version>
             </properties>
-            <build>
```
[the remark this answers is not in the document]
Comment (gerashegalov) [reply pom.xml:697]: yes, thanks, will fix outstanding issues including this.
converted to draft

File: sql-plugin/pom.xml (hunk 1 above)
Comment (gerashegalov) [reply sql-plugin/pom.xml:516]: will redo this part. the idea was to build a dirset for each feature line 32X and then individual patch dirsets.

**Comment by tgravescs** [comment]
before spending more time on this perhaps we should discuss if we want this. I'm a bit torn on it because while it compacts the code I'm not sure it helps readability.

**Comment by gerashegalov** [comment]
> before spending more time on this perhaps we should discuss if we want this. I'm a bit torn on it because while it compacts the code I'm not sure it helps readability.

I cleaned up to demo pattern reuse. Now we are able to express that, for instance 330cdh is like spark33X except don't include nondh

This patterned approach would encourage better naming of the directories to avoid listing exceptions.

**Comment by gerashegalov** [comment]
build

**Comment by gerashegalov** [comment]
build

**Review by jlowe — APPROVED** [review]

**Review by tgravescs — APPROVED** [review]

**Pull request merged**